src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorDiagnosticsEndpoint.cs

```diff
@@ -184,26 +185,56 @@ private static bool ShouldFilterHtmlDiagnosticBasedOnErrorCode(OmniSharpVSDiagno
             return diagnostic.Code.Value.String switch
             {
                 HtmlErrorCodes.UnexpectedEndTagErrorCode => IsHtmlWithBangAndMatchingTags(diagnostic, sourceText, syntaxTree),
-                HtmlErrorCodes.InvalidNestingErrorCode => IsInvalidNestingWarningWithinComponent(diagnostic, sourceText, syntaxTree),
+                HtmlErrorCodes.InvalidNestingErrorCode => IsAnyFilteredInvalidNestingError(diagnostic, sourceText, syntaxTree),
                 HtmlErrorCodes.MissingEndTagErrorCode => FileKinds.IsComponent(syntaxTree.Options.FileKind), // Redundant with RZ9980 in Components
+                HtmlErrorCodes.TooFewElementsErrorCode => IsAnyFilteredTooFewElementsError(diagnostic, sourceText, syntaxTree),
                 _ => false,
             };
 
+            // Ideally this would be solved instead by not emitting the "!" at the HTML backing file,
+            // but we don't currently have a system to accomplish that
+            static bool IsAnyFilteredTooFewElementsError(OmniSharpVSDiagnostic d, SourceText sourceText, RazorSyntaxTree syntaxTree)
+            {
+                var owner = syntaxTree.GetOwner(sourceText, d.Range.Start);
+                var element = owner.FirstAncestorOrSelf<MarkupElementSyntax>(n => n is not null);
+
+                if (!element.StartTag.Name.Content.Equals("html", StringComparison.Ordinal))
+                {
+                    return false;
+                }
+
+                var bodyElement = (MarkupElementSyntax)element.ChildNodes().SingleOrDefault(c => c is MarkupElementSyntax tag && tag.StartTag.Name.Content.Equals("body", StringComparison.Ordinal));
+                return bodyElement is not null && bodyElement.StartTag.Bang is not null;
+            }
+
             // Ideally this would be solved instead by not emitting the "!" at the HTML backing file,
             // but we don't currently have a system to accomplish that
             static bool IsHtmlWithBangAndMatchingTags(OmniSharpVSDiagnostic d, SourceText sourceText, RazorSyntaxTree syntaxTree)
             {
                 var owner = syntaxTree.GetOwner(sourceText, d.Range.Start);
 
-                var element = owner.FirstAncestorOrSelf<MarkupElementSyntax>(n => n is MarkupElementSyntax);
+                var element = owner.FirstAncestorOrSelf<MarkupElementSyntax>(n => n is not null);
                 var startNode = element.StartTag;
                 var endNode = element.EndTag;
+
+                if (startNode is null || endNode is null)
+                {
+                    // If we're missing either the start or end node it's probably because something is unbalanced
```

Comment:
Ah this is the null ref protection piece right? Given that's the case i'd suggest rewording the comment to be more like "we only care about tags with a start and an end because we want to exclude diagnostics from their children" or something to that manner